server/service/schedule/schedule.go

```diff
@@ -0,0 +1,219 @@
+package schedule
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/getsentry/sentry-go"
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/go-kit/log/level"
+)
+
+type schedule struct {
+	ctx        context.Context
+	name       string
+	instanceID string
+	Logger     kitlog.Logger
+
+	muChecks       sync.Mutex
+	configInterval time.Duration
+	schedInterval  time.Duration
+	locker         Locker
+	configCheck    func(start time.Time, wait time.Duration) (*time.Duration, error)
+	preflightCheck func() bool
+
+	muJobs sync.Mutex
+	jobs   map[string]job
+}
+
+func (s *schedule) getConfigInterval() time.Duration {
+	s.muChecks.Lock()
+	defer s.muChecks.Unlock()
+	return s.configInterval
+}
+
+func (s *schedule) setConfigInterval(interval time.Duration) {
+	s.muChecks.Lock()
+	defer s.muChecks.Unlock()
+	s.configInterval = interval
+}
+
+func (s *schedule) getSchedInterval() time.Duration {
+	s.muChecks.Lock()
+	defer s.muChecks.Unlock()
+	return s.schedInterval
+}
+
+func (s *schedule) setSchedInterval(interval time.Duration) {
+	s.muChecks.Lock()
+	defer s.muChecks.Unlock()
+	s.schedInterval = interval
+}
+
+func (s *schedule) SetPreflightCheck(fn func() bool) {
+	s.muChecks.Lock()
+	defer s.muChecks.Unlock()
+	s.preflightCheck = fn
+}
+
+func (s *schedule) SetConfigCheck(fn func(start time.Time, wait time.Duration) (*time.Duration, error)) {
+	s.muChecks.Lock()
+	defer s.muChecks.Unlock()
+	s.configCheck = fn
+}
+
+type job struct {
+	run          func(context.Context) (interface{}, error)
+	statsHandler func(interface{}, error)
+}
+
+type Locker interface {
+	Lock(ctx context.Context, name string, owner string, expiration time.Duration) (bool, error)
+	Unlock(ctx context.Context, name string, owner string) error
+}
+
+func New(ctx context.Context, name string, instanceID string, interval time.Duration, locker Locker, logger kitlog.Logger) (sched *schedule, err error) {
+	sch := &schedule{
+		ctx:        ctx,
+		name:       name,
+		instanceID: instanceID,
+		Logger:     logger,
+
+		configInterval: 1 * time.Hour, // by default we will check for updated config once per hour
+		schedInterval:  interval,
+		locker:         locker,
+
+		jobs: make(map[string]job),
+	}
+	sch.run()
+	return sch, nil
+}
+
+func (s *schedule) AddJob(id string, newJob func(ctx context.Context) (interface{}, error), statsHandler func(interface{}, error)) {
+	s.muJobs.Lock()
+	defer s.muJobs.Unlock()
+	// TODO: guard for job id uniqueness?
+	s.jobs[id] = job{run: newJob, statsHandler: statsHandler}
+}
+
+// each schedule runs in its own go routine
```

Comment:
Is there any mechanism for preventing multiple jobs from running concurrently? I feel like we may want to run them sequentially (eg. to prevent DB utilization spiking extra due to multiple DB-heavy jobs running at the same time).